src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs

```diff
@@ -0,0 +1,170 @@
+            {
+                if (cache.Values.TryGetValue(counter, out var val) ||
+                    (document != null &&
+                     metadataCounters?.BinarySearch(counter, StringComparison.OrdinalIgnoreCase) < 0) ||
+                    cache.GotAll)
+                {
+                    // we either have value in cache,
+                    // or we have the metadata and the counter is not there,
+                    // or GotAll
+
+                    result[counter] = val;
+                    continue;
+                }
+
+                result.Clear();
+                Session.IncrementRequestCount();
+
+                var details = await Session.DocumentStore.Operations.ForDatabase(Session.DatabaseName)
+                    .SendAsync(new GetCountersOperation(DocId, countersArray), token: token).ConfigureAwait(false);
+
+                foreach (var counterDetail in details.Counters)
+                {
+                    cache.Values[counterDetail.CounterName] = counterDetail.TotalValue;
+                    result[counterDetail.CounterName] = counterDetail.TotalValue;
+                }
+
+                break;
+            }
+
+            Session.CountersByDocId[DocId] = cache;
+            return result;
+        }
+
+        public async Task<Dictionary<string, long?>> GetAllAsync(CancellationToken token = default)
+        {
+            if (Session.CountersByDocId.TryGetValue(DocId, out var cache) == false)
+            {
+                cache.Values = new Dictionary<string, long?>(StringComparer.OrdinalIgnoreCase);
+            }
+
+            var missingCounters = cache.GotAll == false;
+
+            if (Session.DocumentsById.TryGetValue(DocId, out var document))
+            {
+                if (document.Metadata.TryGet(Constants.Documents.Metadata.Counters, 
+                    out BlittableJsonReaderArray metadataCounters) == false)
+                {
+                    missingCounters = false;
+                }
+
+                else if (cache.Values.Count >= metadataCounters.Length)
+                {
+                    missingCounters = false;
+                    foreach (var c in metadataCounters)
+                    {
+                        if (cache.Values.ContainsKey(c.ToString()))
+                            continue;
+                        missingCounters = true;
+                        break;
+                    }
+                }
+
+            }
+
+            if (missingCounters)
+            {
+                // we either don't have the document in session and GotAll = false,
+                // or we do and cache doesn't contain all metadata counters
+
+                Session.IncrementRequestCount();
+
+                var details = await Session.DocumentStore.Operations.ForDatabase(Session.DatabaseName)
+                    .SendAsync(new GetCountersOperation(DocId, new string[0]), token: token).ConfigureAwait(false);
```

Comment:
I think it would be better from UX perspective to add a second constructor for this case.

Passing `null` or empty array should not be allowed. Check how it is done here: https://github.com/ravendb/ravendb/blob/43b7e45576d02df20df375eefc5b2da044bfb174/src/Raven.Client/Documents/Commands/GetDocumentsCommand.cs#L30-L51

Reply:
done